### PR TITLE
IDebugSession draft

### PR DESCRIPTION
**Pull request by akaroml**

### CODE REVIEW

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IThreadUtility.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package org.eclipse.jdt.ls.debug;
+
+import com.sun.jdi.ThreadReference;
+import com.sun.jdi.event.StepEvent;
+import java.util.function.Consumer;
+
+import org.eclipse.debug.core.DebugEvent;
+
+interface IThreadUtility {
+    static void stepOver(ThreadReference thread, Consumer<DebugEvent> eventHandler) {
```
Comment (andxu) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IThreadUtility.java]: StepEvent?

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IDebugSession.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    static IDebugSession attach(/* TODO: arguments? */) {
+        throw new UnsupportedOperationException();
+    }
+
+    void detach();
+
+    void terminate();
+
+    // breakpoints
+    void addBreakpoint(String className, int lineNumber, Consumer<IBreakpoint> validatedHandler, Consumer<DebugEvent> breakEventHandler);
```
Comment (andxu) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IDebugSession.java]: I think we can have a BreakPointUtils to convert (String className, int lineNumber,...) into an IBreakpoint?

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IThreadUtility.java (hunk 1 above)
Comment (akaroml) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IThreadUtility.java]: The real event is inside DebugEvent. DebugEvent is to wrap the original event and provide the handler a way of stop resuming by setting its shouldResume to false.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IDebugSession.java (hunk 2 above)
Comment (akaroml) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IDebugSession.java]: To add an breakpoint, you will still need the VM, which is already in DebugSession.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IBreakpoint.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package org.eclipse.jdt.ls.debug;
```
Comment (yaohaizh) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IBreakpoint.java:12]: License header

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IBreakpoint.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+
+import com.sun.jdi.request.EventRequest;
+
+public interface IBreakpoint {
+    // This is the receipt for tracking the breakpoint at JDI side
```
Comment (yaohaizh) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IBreakpoint.java]: For the comment on the public API, it is better to use the javadoc style. /** .. */

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IThreadUtility.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package org.eclipse.jdt.ls.debug;
+
+import com.sun.jdi.ThreadReference;
+import com.sun.jdi.event.StepEvent;
+import java.util.function.Consumer;
+
+import org.eclipse.debug.core.DebugEvent;
+
+interface IThreadUtility {
```
Comment (yaohaizh) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IThreadUtility.java]: If it is utility class, it is not necessary put it in an interface.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/EventHub.java [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package org.eclipse.jdt.ls.debug.internal;
+package org.eclipse.jdt.ls.debug;
```
Comment (testforstephen) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/EventHub.java]: add a new interface IEventHub and keep EventHub implementation in internal package?

**Review by testforstephen — APPROVED** [review]

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IBreakpoint.java (hunk 3 above)
Comment (akaroml) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/IBreakpoint.java:12]: Added. Thanks.

**Pull request merged**